Feature: Pivot Point Indicator

team172011 opened the pull request:
Enhanced #39.

Changes proposed in this pull request:
- Pivot Point Indicators Package: New package with Indicators for calculating standard, Fibonacci and DeMark pivot points and reversals
    - PivotPointIndicator: New indicator for calculating the standard pivot point
        - StandardReversalIndicator: New indicator for calculating the standard reversals (R3,R2,R1,S1,S2,S3)
        - FibonacciReversalIndicator: New indicator for calculating the Fibonacci reversals (R3,R2,R1,S1,S2,S3)
    - DeMarkPivotPointIndicator: New indicator for calculating the DeMark pivot point
        - DeMarkReversalIndicator: new indicator for calculating the DeMark resistance and the DeMark support

- New Testfile:
    - PivotPointIndicatorTest.java

<br>
- [x] added an entry to the unreleased section of `CHANGES.md`

mdeverdelhan commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/FibonacciPivotResistance1Indicator.java:
```diff
@@ -0,0 +1,67 @@
+ * @author team172011(Simon-Justus Wimmer), 09.10.2017
+ * @see <a href="http://stockcharts.com/school/doku.php?id=chart_school:technical_indicators:pivot_points">chart_school: pivotpoints</a>
+ */
+public class FibonacciPivotResistance1Indicator extends RecursiveCachedIndicator<Decimal> {
```
Was it necessary to add FibonacciPivotResistance1Indicator, FibonacciPivotResistance2Indicator, FibonacciPivotResistance3Indicator?
What about only one FibonacciPivotResistanceIndicator with the Fib factor (from an enum, with more precise values) in argument? Same question for FibonacciPivotSupportXXX.

nimo23 commented:
You can do this that way to reduce 6 classes to 3:

package org.ta4j.core.indicators.pivotpoints;

import org.ta4j.core.Decimal;

```
public enum FibonacciFactor {
	
	Factor1(Decimal.valueOf(0.382)), 
	Factor2(Decimal.valueOf(618)), 
	Factor3(Decimal.ONE);

	FibonacciFactor(Decimal factor) {
		this.factor = factor;
	}

	private final Decimal factor;

	public Decimal getFactor() {
		return factor;
	}
}
```


```
package org.ta4j.core.indicators.pivotpoints;

import org.ta4j.core.Decimal;
import org.ta4j.core.Tick;
import org.ta4j.core.indicators.RecursiveCachedIndicator;

import java.util.List;

/**
 * Fibonacci Pivot Resistance Indicator.
 */
public class FibonacciPivotResistanceIndicator extends RecursiveCachedIndicator<Decimal> {

    private PivotPointIndicator pivotPoint;
    private final Decimal fibFactor;

    /**
     * Constructor.
     * <p>
     * Calculates the (fibonacci) resistance reversal
     * @param fibFactor the {@link FibonacciFactor}
     * @param pivotPoint the {@link PivotPointIndicator} for this reversal
     */
    public FibonacciPivotResistanceIndicator(FibonacciFactor fibFactor, PivotPointIndicator pivotPoint) {
        super(pivotPoint);
        this.fibFactor = fibFactor.getFactor();
        this.pivotPoint = pivotPoint;
    }

    @Override
    protected Decimal calculate(int index) {
        List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);
        
        if (ticksOfPreviousPeriod.isEmpty()){
        		return Decimal.NaN;
        }
        
        Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0)); 
        Decimal low = tick.getMinPrice();
        Decimal high =  tick.getMaxPrice();
        
        for(int i: ticksOfPreviousPeriod){
        		Tick t = getTimeSeries().getTick(i);
            low = t.getMinPrice().min(low);
            high = t.getMaxPrice().max(high);
        }
        
        return pivotPoint.getValue(index)
        		.plus(fibFactor.multipliedBy(high.minus(low)));
    }
}
```


```
/**
 * Fibonacci Pivot Support Indicator.
 */
public class FibonacciPivotSupportIndicator extends RecursiveCachedIndicator<Decimal> {

	private PivotPointIndicator pivotPoint;
	private final Decimal fibFactor;

	/**
	 * Constructor.
	 * <p>
	 * Calculates the (fibonacci) support reversal
	 * @param fibFactor the {@link FibonacciFactor}
	 * @param pivotPoint the {@link PivotPointIndicator} for this reversal
	 */
	public FibonacciPivotSupportIndicator(FibonacciFactor fibFactor, PivotPointIndicator pivotPoint) {
		super(pivotPoint);
		this.fibFactor = fibFactor.getFactor();
		this.pivotPoint = pivotPoint;
	}
	
	public FibonacciPivotSupportIndicator(Decimal fibFactor, PivotPointIndicator pivotPoint) {
		super(pivotPoint);
		this.fibFactor = fibFactor;
		this.pivotPoint = pivotPoint;
	}

	@Override
	protected Decimal calculate(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);

		if (ticksOfPreviousPeriod.isEmpty()) {
			return Decimal.NaN;
		}

		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal low = tick.getMinPrice();
		Decimal high = tick.getMaxPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			low = t.getMinPrice().min(low);
			high = t.getMaxPrice().max(high);
		}

		return pivotPoint.getValue(index)
				.minus(fibFactor.multipliedBy(high.minus(low)));
	}
}
```

nimo23 commented:
You can even reduce the 6 classes to only new 1 class 
(which is called **FibonacciPivotPointIndicator**):

```
/**
 * Fibonacci Pivot Point Indicator for Support and Resistance Lines.s
 */
public class FibonacciPivotPointIndicator extends RecursiveCachedIndicator<Decimal> {

	public enum FibonacciFactor {
		Factor1(Decimal.valueOf(0.382)), 
		Factor2(Decimal.valueOf(618)), 
		Factor3(Decimal.ONE);

		FibonacciFactor(Decimal factor) {
			this.factor = factor;
		}
		private final Decimal factor;
	}
	
	public enum FibonacciType {
		Support, 
		Resistance;
	}

    private PivotPointIndicator pivotPoint;
    private final Decimal fibFactor;
    private final FibonacciType fibType;

    /**
     * Constructor.
     * <p>
     * Calculates the (fibonacci) resistance reversal
     * @param pivotPoint the {@link PivotPointIndicator} for this reversal
     * @param fibType the {@link FibonacciType}
     * @param fibFactor the {@link FibonacciFactor}
     */
    public FibonacciPivotResistanceIndicator(PivotPointIndicator pivotPoint, FibonacciType fibType, FibonacciFactor fibFactor) {
        super(pivotPoint);
        this.pivotPoint = pivotPoint;
        this.fibType = fibType;
        this.fibFactor = fibFactor.factor;
    }

	@Override
	protected Decimal calculate(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);

		if (ticksOfPreviousPeriod.isEmpty()) {
			return Decimal.NaN;
		}

		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal low = tick.getMinPrice();
		Decimal high = tick.getMaxPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			low = t.getMinPrice().min(low);
			high = t.getMaxPrice().max(high);
		}

		if (fibType == FibonacciType.Support) {
			return pivotPoint.getValue(index).minus(fibFactor.multipliedBy(high.minus(low)));
		} else {
			// resistance
			return pivotPoint.getValue(index).plus(fibFactor.multipliedBy(high.minus(low)));
		}
	}
}
```

Please use "FibonacciPivotPointIndicator.java" instead of the other 6 classes. Keep it short and simple;)

team172011 commented:
I agree, working with enum would make it easier. I hesitated because it is not so common in other indicators so far. Also the timeLevel id could be defined as enum

nimo23 commented:
For the Standard Pivot Points, you also use 6 redundant classes. This can also be reduced to only one class, in which all its standard pivot point calculations are made (keep them related):
```

import org.ta4j.core.Decimal;
import org.ta4j.core.Tick;
import org.ta4j.core.indicators.RecursiveCachedIndicator;

import java.util.List;

/**
 * Standard Pivot Point Indicator for support and resistance.
 */
public class StandardPivotPointIndicator extends RecursiveCachedIndicator<Decimal> {

	public enum PivotPointType {
		Resistance1, Resistance2, Resistance3, Support1, Support2, Support3;
	}

	private PivotPointIndicator pivotPoint;
	private final PivotPointType pivotPointType;

	/**
	 * Constructor.
	 * <p>
	 * Calculates the (standard) reversal for support or resistance
	 * 
	 * @param pivotPointType the {@link PivotPointType}
	 * @param pivotPointIndicator the {@link PivotPointIndicator}
	 */
	public StandardPivotPointIndicator(PivotPointType pivotPointType, PivotPointIndicator pivotPoint) {
		super(pivotPoint);
		this.pivotPointType = pivotPointType;
		this.pivotPoint = pivotPoint;
	}

	@Override
	protected Decimal calculate(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);

		if (ticksOfPreviousPeriod.isEmpty()) {
			return Decimal.NaN;
		}

		switch (pivotPointType) {
		case Resistance1:
			return calculateResistance1(index);
		case Resistance2:
			return calculateResistance2(index);
		case Resistance3:
			return calculateResistance3(index);
		case Support1:
			return calculateSupport1(index);
		case Support2:
			return calculateSupport2(index);
		case Support3:
			return calculateSupport3(index);
		default:
			return Decimal.NaN;
		}
	}

	private Decimal calculateResistance1(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);
		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal low = tick.getMinPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			low = t.getMinPrice().min(low);
		}

		return Decimal.TWO.multipliedBy(pivotPoint.getValue(index)).minus(low);
	}

	private Decimal calculateResistance2(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);
		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal low = tick.getMinPrice();
		Decimal high = tick.getMaxPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			low = t.getMinPrice().min(low);
			high = t.getMaxPrice().max(high);
		}

		return pivotPoint.getValue(index).plus((high.minus(low)));
	}

	private Decimal calculateResistance3(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);

		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal low = tick.getMinPrice();
		Decimal high = tick.getMaxPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			low = t.getMinPrice().min(low);
			high = t.getMaxPrice().max(high);
		}
		return high.plus(Decimal.TWO.multipliedBy((pivotPoint.getValue(index).minus(low))));
	}

	private Decimal calculateSupport1(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);

		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal high = tick.getMaxPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			high = t.getMaxPrice().max(high);
		}

		return Decimal.TWO.multipliedBy(pivotPoint.getValue(index)).minus(high);
	}

	private Decimal calculateSupport2(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);

		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal high = tick.getMaxPrice();
		Decimal low = tick.getMinPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			high = t.getMaxPrice().max(high);
			low = t.getMinPrice().min(low);
		}
		return pivotPoint.getValue(index).minus((high.minus(low)));
	}

	private Decimal calculateSupport3(int index) {
		List<Integer> ticksOfPreviousPeriod = pivotPoint.getTicksOfPreviousPeriod(index);

		Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
		Decimal high = tick.getMaxPrice();
		Decimal low = tick.getMinPrice();

		for (int i : ticksOfPreviousPeriod) {
			Tick t = getTimeSeries().getTick(i);
			high = t.getMaxPrice().max(high);
			low = t.getMinPrice().min(low);
		}
		return low.minus(Decimal.TWO.multipliedBy((high.minus(pivotPoint.getValue(index)))));
	}
}
```

nimo23 commented:
**So you will end up with only 3 files:**

- PivotPointIndicator.java
- StandardPivotPointIndicator.java
- FibonacciPivotPointIndicator.java

you can add also "DemarkPivotPointIndicator.java".

team172011 commented:
DemarkPivotPointIndicator:

> If Close < Open, then X = High + (2 x Low) + Close

Which close and open? if the **open of current tick** is smaller than the **close of previous period**?

team172011 commented:
Sorry for the vast number of commits. Every time i commit and push via intellij i do not recognice that files are deselected for unknown reason

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/FibonacciReversalIndicator.java:
```diff
@@ -28,24 +28,52 @@
+        SUPPORT_3(Decimal.ONE, false);
+
+        private Decimal fibFactor;
+        private boolean isResistance;
```
No need for "isResistance". Fibonacci Numbers are equal for both resistance and support. Please look at my solution described at https://github.com/ta4j/ta4j/pull/49.

nimo23 replied:
```
public enum FibonacciFactor {
		Factor1(Decimal.valueOf(0.382)), 
		Factor2(Decimal.valueOf(0.618)), 
		Factor3(Decimal.ONE);

		FibonacciFactor(Decimal factor) {
			this.factor = factor;
		}
		private final Decimal factor;
	}
```

team172011 replied:
ah, thought it is always the same -.-

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/FibonacciReversalIndicator.java:
```diff
@@ -0,0 +1,96 @@
+     * Calculates the (fibonacci) support 3 reversal
+     * @param pivotPointIndicator the {@link PivotPointIndicator} for this reversal
+     */
+    public FibonacciReversalIndicator(PivotPointIndicator pivotPointIndicator, PivotLevel level) {
```
Please add a additional constructor with Decimal fibFactor, where user can add another fibonacci-number not listed in enum FibonacciFactor. There exist a lot of more fibonacci numbers also interesting. Look at my solution at https://github.com/ta4j/ta4j/pull/49 for this constructor. By the way, edit your javadoc (it has not all parameters and has old comment).

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/FibonacciReversalIndicator.java:
```diff
@@ -0,0 +1,96 @@
+        if (ticksOfPreviousPeriod.isEmpty())
+            return Decimal.NaN;
+        Decimal high =  getTimeSeries().getTick(ticksOfPreviousPeriod.get(0)).getMaxPrice();
+        Decimal low = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0)).getMinPrice();
```
No need to query for tick again and again.

**Use something like this:**

```
Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0));
Decimal low = tick.getMinPrice();
Decimal high = tick.getMaxPrice();
```

which makes your code easier to maintain and a "little" faster. You only have to change one thing and not 2 or more things when changing anything of the Tick. By the way, it sleeks your code which makes it better readable and better editable.  Look at my solution at https://github.com/ta4j/ta4j/pull/49.

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/FibonacciReversalIndicator.java:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        if (level.isResistance())
+            return pivotPointIndicator.getValue(index).plus(level.getFactor().multipliedBy(high.minus(low)));
```
dont use short if. Use always curly braces. It is better maintainable. Imagine putting anything within if afterwards, then you have to put curly braces. 

Use 

if (level.isResistance()) {
   return pivotPoint.getValue(index).plus(level.getFactor().multipliedBy(high.minus(low)));
}

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/StandardReversalIndicator.java:
```diff
@@ -0,0 +1,137 @@
+
+    private Decimal calculateR3(List<Integer> ticksOfPreviousPeriod, int index){
+        Decimal low = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0)).getMinPrice();
+        Decimal high =  getTimeSeries().getTick(ticksOfPreviousPeriod.get(0)).getMaxPrice();
```
no need to query again and again for tick, put Tick tick = getTimeSeries().getTick(ticksOfPreviousPeriod.get(0)); Look at https://github.com/ta4j/ta4j/pull/49.

nimo23 commented:
> Which close and open? if the open of current tick is smaller than the close of previous period?

**If Close < Open,** 
then X = High(previous day) + [2 x Low(previous day)] + Close(previous day) 

**If Close > Open,** 
then X = [2 x High(previous day)] + Low(previous day) + Close(previous day)

**If Close =** 
Open, then X = High(previous day) + Low(previous day) + [2 x Close(previous day)]


Look at http://www.binarytribune.com/forex-trading-indicators/fibonacci-pivot-points-demark-calculation and http://www.mypivots.com/dictionary/definition/57/demark-pivot-points.

team172011 commented:
I have updated my code and added DeMarkPivotPointIndicator + DeMarkReversalIndicator. The DeMark Pivot Poitn calculation should be correct, but the Support and Resistance values are differ from e.g. tradingview.com but i dont get why.

nimo23 commented:
hmm.. I am in doubt..do you think the "DeMark" indicators are good?  Is it worth to include it? ta4j should not include every indicator, only the "best of the breed". I dont know..

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/DeMarkReversalIndicator.java:
```diff
@@ -0,0 +1,104 @@
+    private final DeMarkPivotPointIndicator pivotPointIndicator;
+    private final DeMarkPivotLevel level;
+
+    public enum DeMarkPivotLevel{
```
Rename this to "PivotType" and put this enum-declaration within "PivotPointIndicator". With this you can reuse "PivotType" also for other "XPivotPointIndicator" (for example, FibonacciPivotIndicator).

team172011 replied:
DeMark Pivot reversals just have one support and one resistance reversal line

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/FibonacciReversalIndicator.java:
```diff
@@ -35,62 +36,46 @@
+    private final FibReversalTyp fibReversalTyp;
+    private final Decimal fibonacciFactor;
 
+    public enum FibReversalTyp{
```
Rename this to "PivotType" or "PivotReversalType" and put this enum-declaration within "PivotPointIndicator". With this you can reuse "PivotType" also for other "XPivotPointIndicator" (for example, DeMarkPivotIndicator). No need to duplicate code.

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/FibonacciReversalIndicator.java:
```diff
@@ -35,62 +36,46 @@
-     * Calculates the (fibonacci) support 3 reversal
+     * Calculates a (fibonacci) reversal
      * @param pivotPointIndicator the {@link PivotPointIndicator} for this reversal
+     * @param fibonacciFactor the fibunacci factor for this reversal (e.g. 1, 0.618, 0.382, ...)
```
add this within this class:
```

public enum FibonacciFactor {
		Factor1(Decimal.valueOf(0.382)), 
		Factor2(Decimal.valueOf(618)), 
		Factor3(Decimal.ONE);

		FibonacciFactor(Decimal factor) {
			this.factor = factor;
		}
		private final Decimal factor;
	}
```

and add a additonal constructor with FibonacciReversalIndicator(FibonacciFactor factor,..) for user convenience. The 3 factors above are the "standard fibonacci factors". Look at my source in https://github.com/ta4j/ta4j/pull/49/ where I add a additional constructor for this.

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/DeMarkPivotPointIndicator.java:
```diff
@@ -0,0 +1,180 @@
+    public static final int PIVOT_TIME_LEVEL_ID_DAY = 1; // 1-, 5-, 10- and 15-minute charts use the prior day's high, low and close.
+    public static final int PIVOT_TIME_LEVEL_ID_WEEK = 2; // 30- 60- and 120-minute charts use the prior week's high, low, and close.
+    public static final int PIVOT_TIME_LEVEL_ID_MONTH = 3; // Pivot Points for daily charts use the prior month's data.
+    public static final int PIVOT_TIME_LEVEL_ID_YEAR = 4; // Pivot Points for weekly and monthly charts use the prior year's data
```
please use enmu:

```
public enum TimeLevel{
tickbased(0), day(1), week(2), month(3), year(4);
..
}

public DeMarkPivotPointIndicator(TimeSeries series, TimeLevel timeLevel){

..}

```

It is also easier to use and type safe:
`DeMarkPivotPointIndicator d = new DeMarkPivotPointIndicator(series, TimeLevel.week);`

instead of

```
// can accept any int-number, which should not be allowed
DeMarkPivotPointIndicator d = new DeMarkPivotPointIndicator(series, 7);
```

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/PivotPointIndicator.java:
```diff
@@ -0,0 +1,165 @@
+public class PivotPointIndicator extends RecursiveCachedIndicator<Decimal> {
+
+    public static final int PIVOT_TIME_LEVEL_ID_TICKBASED = 0;
+    public static final int PIVOT_TIME_LEVEL_ID_DAY = 1; // 1-, 5-, 10- and 15-minute charts use the prior day's high, low and close.
```
replace this with enum:

```
public enum TimeLevel{
tickbased(0), day(1), week(2), month(3), year(4);
..
}

and look at DeMarkPivotIndicator, where you also make use of the "TimeLevel". (dont repeat yourself - DRY;)

nimo23 commented on ta4j/src/main/java/org/ta4j/core/indicators/pivotpoints/PivotPointIndicator.java:
```diff
@@ -0,0 +1,165 @@
+    public static final int PIVOT_TIME_LEVEL_ID_MONTH = 3; // Pivot Points for daily charts use the prior month's data.
+    public static final int PIVOT_TIME_LEVEL_ID_YEAR = 4; // Pivot Points for weekly and monthly charts use the prior year's data
+
+    private final int timeLevel;
```
make it typesafe, replace with enum:

`private final TimeLevel timeLevel;`
